File: project_files/frontlib/model/map.c
/*
 * Hedgewars, a free turn based strategy game
 * Copyright (C) 2012 Simeon Maxein <smaxein@googlemail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include "map.h"

#include "../util/inihelper.h"
#include "../util/util.h"
#include "../util/logging.h"

#include <stdlib.h>

flib_map *flib_map_create_regular(const char *seed, const char *theme, int templateFilter) {
    if(log_badargs_if2(seed==NULL, theme==NULL)) {
        return NULL;
    }
    flib_map newmap = {0};
    newmap.mapgen = MAPGEN_REGULAR;
    newmap.name = "+rnd+";
    newmap.seed = (char*)seed;
    newmap.theme = (char*)theme;
    newmap.templateFilter = templateFilter;
    return flib_map_copy(&newmap);
}

flib_map *flib_map_create_maze(const char *seed, const char *theme, int mazeSize) {
    if(log_badargs_if2(seed==NULL, theme==NULL)) {
        return NULL;
    }
    flib_map newmap = {0};
    newmap.mapgen = MAPGEN_MAZE;
    newmap.name = "+maze+";
    newmap.seed = (char*)seed;
    newmap.theme = (char*)theme;
    newmap.mazeSize = mazeSize;
    return flib_map_copy(&newmap);
}

flib_map *flib_map_create_named(const char *seed, const char *name) {
    if(log_badargs_if2(seed==NULL, name==NULL)) {
        return NULL;
    }
    flib_map newmap = {0};
    newmap.mapgen = MAPGEN_NAMED;
    newmap.name = (char*)name;
    newmap.seed = (char*)seed;
    return flib_map_copy(&newmap);
}

flib_map *flib_map_create_drawn(const char *seed, const char *theme, const uint8_t *drawData, size_t drawDataSize) {
    if(log_badargs_if3(seed==NULL, theme==NULL, drawData==NULL)) {
        return NULL;
    }
    flib_map newmap = {0};
    newmap.mapgen = MAPGEN_DRAWN;
    newmap.name = "+drawn+";
    newmap.seed = (char*)seed;
    newmap.theme = (char*)theme;
    newmap.drawData = (uint8_t*) drawData;
    newmap.drawDataSize = drawDataSize;
    return flib_map_copy(&newmap);
}

flib_map *flib_map_copy(const flib_map *map) {
    flib_map *result = NULL;
    if(map) {
        flib_map *newmap = flib_calloc(1, sizeof(flib_map));
        if(newmap) {
            newmap->mapgen = map->mapgen;
            newmap->drawDataSize = map->drawDataSize;
            newmap->drawData = flib_bufdupnull(map->drawData, map->drawDataSize);
            newmap->mazeSize = map->mazeSize;
            newmap->name = flib_strdupnull(map->name);
            newmap->seed = flib_strdupnull(map->seed);
            newmap->templateFilter = map->templateFilter;
            newmap->theme = flib_strdupnull(map->theme);
            if((newmap->drawData || !map->drawData) && (newmap->name || !map->name) && (newmap->seed || !map->seed) && (newmap->theme || !map->theme)) {
                result = newmap;
                newmap = NULL;
            }
        }
        flib_map_destroy(newmap);
    }
    return result;
}

void flib_map_destroy(flib_map *map) {
    if(map) {
        free(map->seed);
        free(map->drawData);
        free(map->name);
        free(map->theme);
        free(map);
    }
}

File: project_files/Android-build/SDL-android-project/src/org/hedgewars/hedgeroid/Datastructures/Map.java
/*
 * Hedgewars for Android. An Android port of Hedgewars, a free turn based strategy game
 * Copyright (c) 2011-2012 Richard Deurwaarder <xeli@xelification.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

package org.hedgewars.hedgeroid.Datastructures;

import java.io.File;
import java.io.IOException;

import org.hedgewars.hedgeroid.EngineProtocol.EngineProtocolNetwork;


import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Parcel;
import android.os.Parcelable;

public class Map implements Comparable<Map>, Parcelable{

	private static final String MISSION_PREFIX = "Mission: ";

	private String name;
	private String path;
	private String previewPath;
	private MapType type;

	public Map(File mapDir, MapType _type, Context c){
		type = _type;

		name = mapDir.getName();
		path = mapDir.getAbsolutePath();
		previewPath = path + "/preview.png";
		
		/*switch(type){
		case TYPE_DEFAULT:
			
			break;
		case TYPE_GENERATED:
			//TODO
			break;
		case TYPE_MISSION:
			name = MISSION_PREFIX + mapDir.getName();
			path = mapDir.getAbsolutePath();
			break;
		}*/

		
	}
	
	public Map(Parcel in){
		readFromParcel(in);
	}

	public String toString(){
		switch(type){
		default:
		case TYPE_DEFAULT:
			return name;
		case TYPE_GENERATED:
			return "bla";
		case TYPE_MISSION:
			return MISSION_PREFIX + name;
		}
	}
	
	public void sendToEngine(EngineProtocolNetwork epn) throws IOException{
		epn.sendToEngine(String.format("emap %s",name));
	}
	
	public MapType getType(){
		return type;
	}

	public Drawable getDrawable(){
		switch(type){
		case TYPE_MISSION:
		case TYPE_DEFAULT:
			return Drawable.createFromPath(previewPath);
		case TYPE_GENERATED:

		default:
			return null;
		}
	}

	public int compareTo(Map another) {
		switch(type){
		case TYPE_GENERATED:
			switch(another.getType()){
			case TYPE_GENERATED:
				return name.compareTo(another.name);
			case TYPE_MISSION:
				return -1;
			case TYPE_DEFAULT:
				return -1;
			}
		case TYPE_MISSION:
			switch(another.getType()){
			case TYPE_GENERATED:
				return 1;
			case TYPE_MISSION:
				return name.compareTo(another.name);
			case TYPE_DEFAULT:
				return -1;
			}
		case TYPE_DEFAULT:
			switch(another.getType()){
			case TYPE_GENERATED:
				return 1;
			case TYPE_MISSION:
				return 1;
			case TYPE_DEFAULT:
				return name.compareTo(another.name);
			}
		}
		return 0;//default case this should never happen
	}

	public enum MapType{
		TYPE_DEFAULT, TYPE_MISSION, TYPE_GENERATED
	}

	public int describeContents() {
		return 0;
	}
	
	public void writeToParcel(Parcel dest, int flags) {
		dest.writeString(name);
		dest.writeString(path);
		dest.writeString(previewPath);
		dest.writeString(type.name());
	}
	
	private void readFromParcel(Parcel src){
		name = src.readString();
		path = src.readString();
		previewPath = src.readString();
		type = MapType.valueOf(src.readString());
	}
	public static final Parcelable.Creator<Map> CREATOR = new Parcelable.Creator<Map>() {
		public Map createFromParcel(Parcel source) {
			return new Map(source);
		}
		public Map[] newArray(int size) {
			return new Map[size];
		}
		
	};
}
